### PR TITLE
Copy archived js part files out of dart_tool directory

### LINKED ISSUE #34205: [web]: code split / code chunk / deferred import
Similar to https://webpack.js.org/guides/code-splitting/ . I actually found https://news.dartlang.org/2014/08/dart-16-adds-support-for-deferred.html , but it seems Deferred is not yet supported anywhere in Flutter.

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Using deferred imports in Flutter for Web results in additional js part files. These need to be copied out of the build directories and co-located with the rest of the build result. In run mode, we serve these files when requested (only for release and profile).

This adds some usage of the archive package, but this is already a dependency and is included in the BUILD.gn file already.

Fixes https://github.com/flutter/flutter/issues/34205

### CODE REVIEW

File: packages/flutter_tools/lib/src/build_runner/build_script.dart [hunk 1]
```diff
@@ -461,8 +463,32 @@ Future<void> bootstrapDart2Js(BuildStep buildStep, String flutterWebSdk, bool pr
   final AssetId jsOutputId = dartEntrypointId.changeExtension(jsEntrypointExtension);
   final File jsOutputFile = scratchSpace.fileFor(jsOutputId);
   if (result.succeeded && jsOutputFile.existsSync()) {
-    log.info(result.output);
-    // Explicitly write out the original js file and sourcemap.
+    final String rootDir = path.dirname(jsOutputFile.path);
+    final String dartFile = path.basename(dartEntrypointId.path);
+    final Glob fileGlob = Glob('$dartFile.js*');
+    final Archive archive = Archive();
+    await for (FileSystemEntity jsFile in fileGlob.list(root: rootDir)) {
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/build_runner/build_script.dart]: `.whereType<File>()` would remove the need for extra type checks inside the loop.

File: packages/flutter_tools/lib/src/build_runner/web_fs.dart [hunk 2]
```diff
@@ -175,21 +174,23 @@ class WebFs {
...
-        '${targetBaseName}_web_entrypoint.dart.bootstrap.js',
+          '${targetBaseName}_web_entrypoint.dart.bootstrap.js',
       '${targetBaseName}_web_entrypoint.digests': 'packages/${flutterProject.manifest.appName}/'
-        '${targetBaseName}_web_entrypoint.digests',
+          '${targetBaseName}_web_entrypoint.digests',
     };
+
+    // Initialize the dwds server.
+    final int hostPort = port == null ? await os.findFreePort() : int.tryParse(port);
+    // Map the bootstrap files to the correct package directory.
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/build_runner/web_fs.dart]: Move this comment one line down?

File: packages/flutter_tools/lib/src/build_runner/web_fs.dart [hunk 3]
```diff
@@ -259,6 +261,32 @@ class WebFs {
...
+        // entrypoint + a "part" suffix (Though the actual names are arbitrary).
+        // To make this easier to deal with they are copied into a temp directory.
+        if (partFiles == null) {
+          final File dart2jsArchive = fs.file(fs.path.join(
+              flutterProject.dartTool.path,
+              'build',
+              'flutter_web',
+              '${flutterProject.manifest.appName}',
+              'lib',
+              '${targetBaseName}_web_entrypoint.dart.js.tar.gz'));
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/build_runner/web_fs.dart]: nit: a trailing comma would compel dartfmt to for this a bit better.

File: packages/flutter_tools/lib/src/build_runner/web_fs.dart [hunk 4]
```diff
@@ -259,6 +261,32 @@ class WebFs {
...
+          if (dart2jsArchive.existsSync()) {
+            final Archive archive = TarDecoder().decodeBytes(dart2jsArchive.readAsBytesSync());
+            partFiles = fs.systemTempDirectory.createTempSync('_flutter_tool')
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/build_runner/web_fs.dart]: Can this overflow temp disk space if not cleaned up explicitly?

File: packages/flutter_tools/lib/src/web/compile.dart [hunk 5]
```diff
@@ -48,9 +49,22 @@ Future<void> buildWeb(FlutterProject flutterProject, String target, BuildInfo bu
         flutterProject.manifest.appName,
         '${fs.path.withoutExtension(target)}_web_entrypoint.dart.js',
       );
+      // Check for deferred import outputs.
+      final File dart2jsArchive = fs.file(fs.path.join(
+          flutterProject.dartTool.path,
+          'build',
+          'flutter_web',
+          '${flutterProject.manifest.appName}',
+          '${fs.path.withoutExtension(target)}_web_entrypoint.dart.js.tar.gz'));
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/web/compile.dart]: nit: trailing comma for better readability?

**Review by yjbanov — APPROVED** [review]

File: packages/flutter_tools/lib/src/build_runner/build_script.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_runner/build_script.dart]: Interestingly `Stream` doesn't define a `whereType`

File: packages/flutter_tools/lib/src/web/compile.dart (hunk 5 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/web/compile.dart]: Fixed

File: packages/flutter_tools/lib/src/build_runner/web_fs.dart (hunk 4 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_runner/web_fs.dart]: Made this into a class with a dispose method

File: packages/flutter_tools/lib/src/build_runner/web_fs.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_runner/web_fs.dart]: Done

File: packages/flutter_tools/lib/src/build_runner/web_fs.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_runner/web_fs.dart]: Removed, it was outdated and unclear

**Comment by tvolkert** [comment]
This PR has been implicated in a benchmark regression -- https://github.com/flutter/flutter/issues/41392

**Pull request merged**